socksio/singledispatchmethod.py

```diff
@@ -0,0 +1,42 @@
+"""Backport of @functools.singledispatchmethod to Python <3.7.
+
+Adapted from https://github.com/ikalnytskyi/singledispatchmethod
+removing 2.7 specific code.
+"""
+
+import functools
+import sys
+
+
+if sys.version_info >= (3, 8):
+    singledispatchmethod = functools.singledispatchmethod
+else:
+    update_wrapper = functools.update_wrapper
+    singledispatch = functools.singledispatch
+
+    class singledispatchmethod(object):
```

Comment:
We'll need types on these, totally fine to sprinkle `typing.Any` and `# type: ignore` if needed. :)